subprojects/dependency-management/src/main/java/org/gradle/internal/resolve/resolver/DefaultArtifactSelector.java

```diff
@@ -99,31 +102,32 @@ private ResolvedVariant toResolvedVariant(VariantResolveMetadata.Identifier iden
                                               ExcludeSpec exclusions,
                                               ModuleVersionIdentifier ownerId,
                                               ModuleSources moduleSources,
-                                              @Nullable Map<VariantResolveMetadata.Identifier, ResolvedVariant> resolvedVariantCache) {
+                                              ResolvedVariantCache resolvedVariantCache,
+                                              boolean eligibleForCaching) {
         // artifactsToResolve are those not excluded by their owning module
         List<? extends ComponentArtifactMetadata> artifactsToResolve = CollectionUtils.filter(artifacts,
                 artifact -> !exclusions.excludesArtifact(ownerId.getModule(), artifact.getName())
         );
 
         boolean hasExcludedArtifact = artifactsToResolve.size() < artifacts.size();
+        ImmutableAttributes attributes = artifactTypeRegistry.mapAttributesFor(variantAttributes, artifacts);
 
-        if (hasExcludedArtifact || identifier == null) {
+        if (hasExcludedArtifact) {
             // An ad hoc variant, has no identifier
```

Comment:
🧐 If it has no identifier, does this mean the `identifier` param is null and should be marked `@Nullable`?